File: D3/kMC/Ex02/Ex02.f90

  module variables
    use iso_c_binding
    implicit none
    !
    integer( c_int ), dimension(:), pointer :: site, nneig, neig, nevt, spec, event_site
    !
    real( c_double ), dimension(:), pointer :: rate, prop, event_rate
    !
    ! ::: EVENT
    integer( c_int ), dimension(:), pointer :: init_state, final_state
    real( c_double ), dimension(:), pointer :: ebarrier, de, f0
    real( c_double ), dimension(:,:), pointer :: ebond
    !
    ! ::: TIME
    real( c_double ) :: t_choose
    !
    ! ::: Local update
    integer( c_int ) :: save_site, save_gp(25)
    !
    ! ::: Brownian Mvt
    integer( c_int ) :: navg
    integer( c_int ), dimension(3) :: rold, pos
    !
  end module variables

!
!  USER routine:
!  - subroutine read_event( obj ) 
!  - subroutine event_rate_calc( obj )
!  - subroutine 
!
! .................................................................................................

    subroutine read_event( struc ) bind( C )
      use iso_c_binding
      use derived_types
      use sub_new_types
      use errors
      use variables
      implicit none
      !
      type( KMC_type ), intent( inout ) :: struc
      character (len=500,kind=c_char)    :: string
      integer( c_int )                  :: u0, i, id, ibd, jbd, ios, nevent, nbond
      logical                           :: EOF
      !
      character (len=1,kind=c_char)  :: delims = " "
      CHARACTER (len=100,kind=c_char), dimension (50) :: args
      integer( c_int ) :: nargs
      !
      !  ::: INIT save site
      !
      save_site = 0 
      rold = 0
      navg = 0
      t_choose = 0.0
      !
      !  ::: Lecture of state and rate of each node
      !
      print*, " INPUT_EVENT: ",trim(struc% input_event)
      open( newunit=u0, file=trim(struc% input_event), iostat=ios )
      if ( ios /= 0 ) call error( "input_event does't open!!" )
      !
      EOF = .false.
      do while ( .not.EOF )
         !
         call read_line( u0, string, EOF )
         call parse( trim(string), delims, args, nargs )
         !
         write (*,*) "Lu :", nargs, (i,trim(args(i)), i=1,nargs)
         !
         if ( args(1) == "Number_of_event" ) then
            read( args(2), '(i5)' ) nevent
            call builder_event_type( struc% event, nevent )
            !
            call link_int1_ptr( struc% event% ptr_i_state, init_state, struc% event% nevent)
            call link_int1_ptr( struc% event% ptr_f_state, final_state, struc% event% nevent)
            call link_real1_ptr( struc% event% ptr_f0, f0, struc% event% nevent)
            call link_real1_ptr( struc% event% ptr_ebarrier, ebarrier, struc% event% nevent)
            call link_real1_ptr( struc% event% ptr_de, de, struc% event% nevent)
            !
            do i = 1,struc% event% nevent
               read (u0,*) id, init_state(id), final_state(id), f0(id), ebarrier(id), de(id)
               write (*,*) id, init_state(id), final_state(id), f0(id), ebarrier(id), de(id)
            enddo
            !
         endif
         !
         !
         if ( args(1) == "Energy_Bond" ) then
            !
            read( args(2), '(i5)' ) nbond
            !
            if ( nbond /= 0 ) then
               !
               call link_real2_ptr( struc% event% ptr_ebond, ebond, struc% event% nbond, struc% event% nbond)
               !
               do i = 1,nbond
                  call read_line( u0, string, EOF )
                  call parse( trim(string), delims, args, nargs )
                  read( args(2), '(i5)') ibd
                  read( args(3), '(i5)') jbd
                  read( args(4), '(f10.5)') ebond(ibd,jbd)
                  write (*,*) i, ibd, jbd, ebond( ibd, jbd )
               enddo
            endif
            !
         endif
         !
      enddo
      !
      !
      close( u0 )
      !
    end subroutine read_event
! .................................................................................................

    subroutine event_rate_calc( struc ) bind( C )
      use iso_c_binding
      use derived_types
      use sub_new_types
      use errors
      use variables
      implicit none

      type( KMC_type ), intent( inout ) :: struc
      integer( c_int ) ::  is 

      call link_int1_ptr( struc% ptr_site,             site,       struc% tot_sites )
      call link_int1_ptr( struc% ptr_nneig,            nneig,      struc% tot_sites )
      call link_int1_ptr( struc% ptr_neig,             neig,       nvois*struc% tot_sites )
      call link_real1_ptr( struc% ptr_rate,            rate,       struc% tot_sites )
      call link_real1_ptr( struc% ptr_event_rate,      event_rate, nvois*struc% tot_sites )
      !
      call link_real1_ptr( struc% event% ptr_f0,       f0,         struc% event% nevent )
      call link_real1_ptr( struc% event% ptr_ebarrier, ebarrier,   struc% event% nevent )
      call link_real1_ptr( struc% event% ptr_de,       de,         struc% event% nevent )
      call link_real2_ptr( struc% event% ptr_ebond,    ebond,      struc% event% nbond, struc% event% nbond )
      !
      do is = 1,obj% tot_sites   

      enddo
      !
      !
      !
    end subroutine event_rate_calc
! ..................................................................................................

    subroutine choose_event( struc, isite, ievent ) bind( C )
      use iso_c_binding
      use derived_types
      use sub_new_types
      use random
      use variables
      implicit none

      type( KMC_type ), intent( inout ) :: struc
      integer( c_int ), intent( inout ) :: isite, ievent

      integer( c_int ) :: i, j0, jn
      real( c_double ) :: rsum, rdn, rrdn, t_start, t_stop

      call link_int1_ptr( struc% ptr_site,        site,       struc% tot_sites )
      call link_int1_ptr( struc% ptr_nneig,       nneig,      struc% tot_sites )
      call link_int1_ptr( struc% ptr_neig,        neig,       nvois*struc% tot_sites )
      call link_real1_ptr( struc% ptr_event_rate, event_rate, nvois*struc% tot_sites )
      call link_real1_ptr( struc% ptr_rate,       rate,       struc% tot_sites )
      !
      !
      isite = 0 ; ievent = 0
      call random_number( rdn )
      rrdn = rdn*struc% sum_rate
      struc% rand_rate = rrdn
      !
      call cpu_time( t_start )
      !      
      rsum = 0.0
      do i = 1,struc% tot_sites
         !
         j0 = nneig( i )
         do jn = 1,neig( j0 )
         !  !
            ievent = jn
            rsum = rsum + event_rate( j0 + jn )
            if ( rsum > rrdn ) exit
         !  !
         enddo
         !
         isite = i
         if ( rsum > rrdn ) exit
      enddo
      !
      if ( rsum <= rrdn ) then
         write (*,*) " PB choose event...", rsum, rrdn, struc% sum_rate
         stop " EXIT..."
      endif
      !
      call cpu_time( t_stop )
      !
      t_choose = t_choose + t_stop - t_start
      !
    end subroutine choose_event
! ..................................................................................................
    
    subroutine event_applied( struc, is, jn ) bind( C )
      use iso_c_binding
      use derived_types
      use sub_new_types
      use variables
      implicit none
      !
      type( KMC_type )               :: struc
      integer( c_int ), intent( in ) :: is, jn
      integer( c_int )               :: j, jv, j0, k, k0, kv,l,  &
                                        igp, see, nx, ny, nz, nxy
      !
      integer( c_int ), dimension(3) :: x
      !
      !
      call link_int1_ptr( struc% ptr_site, site, struc% tot_sites )
      call link_int1_ptr( struc% ptr_nneig, nneig, struc% tot_sites )
      call link_int1_ptr( struc% ptr_neig, neig, nvois*struc% tot_sites )
      call link_real1_ptr( struc% ptr_rate, rate, struc% tot_sites )
      call link_real1_ptr( struc% ptr_event_rate, event_rate, nvois*struc% tot_sites )
      !
      !  ::: Flip 0 -> 1
      !
      j0 = nneig( is )
      if ( site( is ) == 1 ) then
         call warning( " site Flip Problem 0 = 1 ")
         write (*,*) is, site( is ), ( neig(j0+j), site( neig(j0+j) ),j=1,neig(j0) )
         write (*,*) is, rate( is ), ( neig(j0+j), event_rate( j0+j ),j=1,neig(j0) )
      endif
      !
      site( is ) = 1
      !
      ! ::: Empty the table
      rate( is ) = 0.0
      do jv = 1,neig( j0 )
         event_rate( j0 + jv ) = 0.0   
      enddo
      !
      j = neig( j0 + jn )
      if ( site( j ) == 0 ) call warning( " site Flip Problem 1 = 0 ")
      !
      site( j ) = 0
      !
      !
    end subroutine event_applied    
! ..................................................................................................
!
    subroutine analyse( obj ) bind( C )
      use iso_c_binding
      use derived_types
      use sub_new_types
      use variables
      implicit none
      !
      type( KMC_type ) :: obj
      integer( c_int ) :: i, j, j0, jv, k, k0, kv,   &
                          ngp, gpv, nc, mixgp, nvac, &
                          maxsize, max_gp, nx, ny, nz, nxy
      real( c_double ) :: x, y, z, t_start, t_stop
      !
      integer( c_int ), dimension( obj% tot_sites ) :: gp, histo
      integer( c_int ), dimension( -1:int(obj% tot_sites/2) ) :: clster
      !
      call link_int1_ptr( obj% ptr_site, site, obj% tot_sites )
      call link_int1_ptr( obj% ptr_nneig, nneig, obj% tot_sites )
      call link_int1_ptr( obj% ptr_neig, neig, nvois*obj% tot_sites )
      call link_real1_ptr( obj% ptr_prop, prop, obj% nprop )
      !
      !
      call cpu_time( t_start )
      !
      ! ::: Cluster size :::
      !
      ngp = 0 ; gp = 0 ; max_gp = 0
      clster = 0 ; nvac = 0
      do i = 1,obj% tot_sites
         !
         if ( site( i ) == 1 ) cycle
         !
         nvac = nvac + 1
         nc = 0
         gp( i ) = -1
         gpv = 0 ; mixgp = 0
         j0 = nneig( i )
         do jv = 1,neig( j0 )
            j = neig( j0 + jv )
            !
            if ( site( j ) == 1 ) cycle
            !
            if ( gp( j ) /= 0.and.gpv == 0 ) then
               gpv = gp( j )
            elseif ( gp( j ) /= 0.and.gpv /= 0.and.gp(j) /= gpv ) then
               mixgp = gp( j )
            endif
            !
            nc = nc + 1
            !
         enddo
         !
         !write (*,*) i,nc,ngp,gpv,mixgp
         !
         if ( nc /= 0.and.gpv /= 0.and.mixgp == 0 ) then
            gp( i ) = gpv
         elseif ( nc /= 0.and.gpv == 0 ) then
            ngp = ngp + 1
            gp( i ) = ngp
            do jv = 1,neig( j0 )
               j = neig( j0 + jv )
               if ( site( j ) == 0 ) gp( j ) = gp( i )
            enddo
         elseif ( mixgp /= 0 ) then
            !
            gp( i ) = min( mixgp, gpv )
            do j = 1,i-1
               if ( site( j ) == 1 ) cycle
               if ( gp( j ) == mixgp.or.gp( j ) == gpv ) then
                  gp( j ) = gp( i )
                  k0 = nneig( j )
                  do kv = 1,neig( k0 )
                     k = neig( k0 + kv )
                     if ( site( k ) == 0 ) gp( k ) = gp( i )
                  enddo
               endif
            enddo
            !
         endif
         !
      enddo
      !
      clster = 0
      max_gp = 0
      do i = 1,obj% tot_sites
         clster( gp(i) ) = clster( gp(i) ) + 1
         max_gp = max( max_gp, gp(i) )
      enddo
      !
      nvac = clster(-1); nc = 0
      histo = 0 ; maxsize = 1
      histo( 1 ) = clster( -1 )
      !
      do i = 1,max_gp
         nvac = nvac + clster( i )
         if ( clster(i) /= 0 ) nc = nc + 1
         histo( clster(i) ) = histo( clster(i) ) + 1
         maxsize = max( maxsize, clster(i) )
      enddo
      !
      prop( 1 ) = 0
      do i = 1,maxsize
         prop( 1 ) = prop( 1 ) + real(i*histo( i ))/real( clster(-1) + nc )
      enddo
      !
      !
      call cpu_time( t_stop )
      !
      !
    end subroutine analyse
! ..................................................................................................
! ..................................................................................................

File: D3/kMC/Ex02/Ex02.c
#include <string.h>
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <time.h>

/* ------- VARIABLE DECLARATION... 
   ------------------------------- */

// ::: PHYSICAL PARAMETER
double kb = 1.38065e-23; 
double kbev = 8.61733e-5;
double na = 6.02214e23;
double pi = 4.0*atan(1.0);
double u = 1.66e-27;
double ang = 1e-10;

// ::: STRUCTURE
struct event_type {

  int nevent, nbond, nchem_react;

  int *ptr_i_state,   // array contains Initial state 
      *ptr_f_state;   // array contains Final state 

  double *ptr_f0,          // array contains prefactor rate 
         *ptr_ebarrier,    // array contains activation energies
         *ptr_de,          // array contains difference energy between two miniums
         **ptr_ebond;      // 2D array contains bond energies
};

struct kmc_type {

  int bavard, conv;          // Verbose & convergence
  int period[3], nsites[3];  // Bound periodic condition & number of site 

  char algorithm[50],        // algorithm name: BKL or Gillepsie
       input_file[50],       // input_file name
       input_event[50],      // event file name
       libname[50],          // shared lib name with the path
       init_mod[50];         // Initialization mode: species, random, none 

  int tot_sites,             // Total number of sites
      max_step,              // max step criterium
      sys_dim,               // system dimension, 1, 2 or 3 D
      freq_write,            // write stat frequency
      nprop,                 // number of properties compute in analyse routine
      node_state,            // number of state of one site
      nspec,                 // number of different sepcies
      npressure,             // number of partial pressure
      step;                  // actual step

  double sum_rate,           // Total rate of system 
         rand_rate,          // random rate 
         time,               // actual time
         rand_time,          // random time
         max_time,           // max time criterium
         temp,               // temperature value (Kelvin)
         kt,                 // Energy of temperature
         per100,             // Percentage of 0 state in initial system
         f0,                 // prefactor rate value
         scale;              // scale value 

  int     *ptr_site, *ptr_nneig, *ptr_nevt, *ptr_neig, *ptr_event_site, *ptr_spec;
  double  *ptr_rate, *ptr_prop, *ptr_event_rate, *ptr_pressure, *ptr_masse;

  struct event_type event;
};


extern int *__hidden_table_MOD_h_i_state, *__hidden_table_MOD_h_f_state;
extern double *__hidden_table_MOD_h_f0, *__hidden_table_MOD_h_ebarrier, *__hidden_table_MOD_h_de, **__hidden_table_MOD_h_ebond;


/* ------- FUNCTION DECLARATION...
   ------------------------------- */

void builder_event_type( struct event_type *, int * );


/* ------- FUNCTION DEFINITION... 
   ------------------------------ */

char ** parsing( char *string, int len ) {
  int i = 0;
  char **pch;
  pch = malloc( len*sizeof(char) );
  pch[i] = strtok( string, "\t " );
  while ( pch[i] != NULL ) {
   // printf( "sparse %d %s \n",i, pch[i] );
    i++;
    pch[i] = strtok( NULL, "\t " );
  }
  return pch;
}
// ...........................................................

// ......................................................................

double random_number( double min, double max ) {
   return min + (double)rand() / ( (double)(RAND_MAX + 1.0)*( max - min) );
}
// ......................................................................
// ......................................................................

void builder_event_type( struct event_type *event, int *nevent ) {
  int int_size = sizeof( int );
  int double_size = sizeof( double );

  printf( " Enter in EVENT_type constructor...\n" );
  event->nevent = *nevent;

  __hidden_table_MOD_h_i_state = malloc( *nevent*int_size );
  __hidden_table_MOD_h_f_state = malloc( *nevent*int_size );
  __hidden_table_MOD_h_f0 = malloc( *nevent*double_size );
  __hidden_table_MOD_h_ebarrier = malloc( *nevent*double_size );
  __hidden_table_MOD_h_de = malloc( *nevent*double_size );

  __hidden_table_MOD_h_ebond = malloc( event->nbond*sizeof( double* ) );
  for (int i = 0; i < event->nbond; i++ ) 
    __hidden_table_MOD_h_ebond[i] = malloc( event->nbond*sizeof( double** ) );

  event->ptr_i_state  = __hidden_table_MOD_h_i_state;
  event->ptr_f_state  = __hidden_table_MOD_h_f_state;
  event->ptr_ebarrier = __hidden_table_MOD_h_ebarrier;
  event->ptr_f0       = __hidden_table_MOD_h_f0; 
  event->ptr_de       = __hidden_table_MOD_h_de;
  event->ptr_ebond    = __hidden_table_MOD_h_ebond;

  printf( " EVENT_type constructor DONE \n" );

}
// ...........................................................

void read_event( struct kmc_type *struc ) {

  char string[100], **word;
  char *b = string;
  int i, id, nevent, nbond, ibd, jbd;
  size_t bufsize = 100;
  size_t nword;

  FILE *fp;

  fp = fopen( struc->input_event, "r" );
  if ( fp == NULL ) perror( " PB open input_event...\n" );

  nword = getline(&b, &bufsize, fp );

  while ( strcmp(&string[0],"#") >= 0 ) {

     word = parsing( string, bufsize );

     if ( !strcmp(word[0],"Number_of_event") ) {
        nevent = atoi( word[1] );
        printf( " nevent %d\n ",nevent );
        builder_event_type( &struc->event, &nevent );

        /* ==== read the event ==== */
        nword = getline(&b, &bufsize, fp );
        int n = struc->event.nevent ;
        for ( id = 0; id < struc->event.nevent; id ++ ) {
            word = parsing( string, bufsize );

            struc->event.ptr_i_state[ id ]  = atoi( word[1] );
            struc->event.ptr_f_state[ id ]  = atoi( word[2] );
            struc->event.ptr_f0[ id ]       = atof( word[3] );
            struc->event.ptr_ebarrier[ id ] = atof( word[4] );
            struc->event.ptr_de[ id ]       = atof( word[5] );
            printf( " read %d event %d %d %f %f \n", id, struc->event.ptr_i_state[ id ], struc->event.ptr_f_state[ id ],
                  struc->event.ptr_de[ id ], struc->event.ptr_ebarrier[ id ], struc->event.ptr_de[ id ] );

            nword = getline(&b, &bufsize, fp );
         }

     }

     if( !strcmp( word[0], "Energy_Bond") ) {
         nbond = atoi( word[1] );
         printf( " dans energy_bond %s %d \n",word[0],nbond );

         /* ==== read Energy Bond ==== */
         for ( id = 0; id < nbond; id++ ) {
             nword = getline(&b, &bufsize, fp );
             word = parsing( string, bufsize );
             ibd = atoi( word[1] );
             jbd = atoi( word[2] );
             struc->event.ptr_ebond[ibd][jbd] = atof( word[3] );
             //printf( " %d %d %f %f --\n", ibd, jbd, struc->event.ptr_ebond[ibd][jbd] );
         }

      }

     nword = getline(&b, &bufsize, fp );

  } // --- WHILE ----


  fclose( fp );

  srand( time(NULL) );
 // exit(1);

}
// .................................................................................................

void event_rate_calc( struct kmc_type *struc ) {


}
// ....................................................................................
// ..................................................................................................

void choose_event( struct kmc_type *struc, int *isite, int *ievent ) {
   
   int i, j0, jn, j, nvj;

   *isite = 0;
   *ievent = 0;
   
   double rdn = random_number( 0.0, 1.0 );
   double rrdn = rdn*struc->sum_rate;
   double *evt_rate = struc->ptr_event_rate;
   struc-> rand_rate = rrdn;
   double rsum = 0.0;

   for (i = 0; i < struc->tot_sites; i++) {
      
      j0 = struc->ptr_nneig[ i ] - 1;
      nvj = struc->ptr_neig[ j0 ];

      for ( jn = 1; jn <= nvj; jn++) {

          *ievent = jn;
          rsum += evt_rate[ j0 + jn ];
          if ( rsum > rrdn ) break;

      }

      *isite = i;
      if ( rsum > rrdn ) break;

   }

   if ( rsum <= rrdn ) 
     printf( " PB choose_event... %f %f %f\n", rrdn, rsum, struc->sum_rate);
}
// ..................................................................................................

void event_applied( struct kmc_type *struc, int *is, int *jn ) {

  int j, j0;
  
  if ( struc->ptr_site[ *is ] == 1 ) {
    printf( " site Flip Problem 0 = 1 : %d %d \n", *is, struc->ptr_site[ *is ]);
  }

  struc->ptr_site[ *is ] = 1;
  j0 = struc->ptr_nneig[ *is ] - 1;
  j = struc->ptr_neig[ j0 + *jn ] - 1;

  if ( struc->ptr_site[ j ] == 0 ) 
    printf( " site Flip Problem 1 = 0 : %d %d \n", j, struc->ptr_site[ j ]);

  struc->ptr_site[ j ] = 0  ;
}
// ..................................................................................................

void analyse( struct kmc_type *obj ) {

  int i,j, j0, jv, jn,k0, k, kv, nvj, nvk, ngp, gpv, nc, mixgp, nvac, maxsize, max_gp;

  int gp[obj->tot_sites], histo[obj->tot_sites];
  int clster[obj->tot_sites];

  for ( i = 0; i < obj->tot_sites; i++) {
    clster[ i ] = 0;
    gp[ i ] = 0;
  }
/* ----------------------------------------
   1) We identify the cluster:
      gp = 0        : no already treated
      gp = 1        : size = 1
      gp = {2,..,N} : size > 1
   2) We count the size of cluster 
   ---------------------------------------- */

  ngp = 1; max_gp = 0; nvac = 0;
  for ( i = 0; i < obj->tot_sites; i++ ) {

      if ( obj->ptr_site[ i ] == 1 ) continue;

      nvac += 1;
      nc = 0;
      gp[ i ] = 1;
      gpv = 0; mixgp = 0;

      j0 = obj->ptr_nneig[ i ] - 1;
      nvj = obj->ptr_neig[ j0 ];

      for ( jn = 1; jn <= nvj; jn++ ) {

          j = obj->ptr_neig[ j0 + jn ] - 1;
          
          if ( obj->ptr_site[ j ] == 1 ) continue; 

          if ( gp[ j ] != 0 && gpv == 0 ) {
             gpv = gp[ j ];
          } else if ( gp[ j ] != 0 && gpv != 0 && gp[ j ] != gpv ) {
             mixgp = gp[ j ];
          }

          nc += 1;

      } // ----

      if ( nc != 0 && gpv != 0 && mixgp == 0 ) {

         gp[ i ] = gpv;

      } else if ( nc != 0 && gpv == 0 ) {

         ngp += 1;
         gp[ i ] = ngp;

         j0 = obj->ptr_nneig[ i ] - 1;
         nvj = obj->ptr_neig[ j0 ];

         for ( jv = 1; jv <= nvj; jv++ ) {
             j = obj->ptr_neig[ j0 + jv ] - 1;
             if ( obj->ptr_site[ j ] == 0 ) gp[ j ] = gp[ i ];
         }

      } else if ( mixgp != 0 ) {

         gp[ i ] = (int)fmin( mixgp, gpv );

         for ( j = 0; j < i; j++ ) {

             if ( obj->ptr_site[ j ] == 1 ) continue;

             if ( gp[ j ] == mixgp || gp[ j ] == gpv ) {
                gp[ j ] = gp[ i ];

                k0 = obj->ptr_nneig[ j ] - 1;
		nvk = obj->ptr_neig[ k0 ];

                for ( kv = 1; kv <= nvk; kv++ ) {
		    k = obj->ptr_neig[ k0 + kv ] - 1;
                    if ( obj->ptr_site[ k ] == 0 ) gp[ k ] = gp[ i ];
                } 

             }
         } 

      } // --- else if

  } // ---- for "i"

  for ( i = 0; i < obj->tot_sites; i++ ) {
      clster[ gp[ i ] ] += 1;
      max_gp = fmax( max_gp, gp[ i ] );
      histo[ i ] = 0;
  }

  nvac = 0;  nc = 0;
  maxsize = 1;
  for ( i = 0; i < max_gp; i++ ) {
      nvac += clster[ i ];
      if ( clster[i] != 0 ) nc += 1;
      histo[ clster[i] ] += 1;
      maxsize = (int)fmax( maxsize, clster[i] ); 
  }
  obj->ptr_prop[ 0 ] = 0.;
  for ( i = 0; i < maxsize; i++ ) {
      obj->ptr_prop[ 0 ] += ((double)( (i+1)*histo[i] ))/((double)( clster[1] + nc ));
  }
}

//! ..................................................................................................






